File: app/src/main/java/com/jdappel/beerinvestigator/ui/viewmodel/impl/BeerViewModelImpl.java
package com.jdappel.beerinvestigator.ui.viewmodel.impl;

import com.google.common.collect.Lists;
import com.jdappel.beerinvestigator.rest.Beer;
import com.jdappel.beerinvestigator.rest.BreweryDBApi;
import com.jdappel.beerinvestigator.rest.BreweryDBResponse;
import com.jdappel.beerinvestigator.ui.viewmodel.BeerViewModel;

import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import rx.Observable;
import rx.subjects.BehaviorSubject;
import rx.subscriptions.CompositeSubscription;

/**
 * Implements {@link com.jdappel.beerinvestigator.ui.viewmodel.BeerViewModel} to retrieve a list of
 * beers based on the search criteria and populate the view.
 */
class BeerViewModelImpl implements BeerViewModel {

    private final BreweryDBApi beerService;
    private final BehaviorSubject<List<Beer>> subject = BehaviorSubject.create();
    private final CompositeSubscription subscriptions = new CompositeSubscription();

    @Inject
    BeerViewModelImpl(final BreweryDBApi beerService) {
        this.beerService = beerService;
    }

    @Override public void subscribe(Observable<String> searchString, Observable<Boolean> checkbox) {
        Observable<List<Beer>> beers = searchString.flatMap(query -> beerService.getBeers(query)
                .filter(obs -> obs.getData() != null && !obs.getData().isEmpty())
                .map(BreweryDBResponse::getData));

        Observable<List<Beer>> finalList =
                Observable.combineLatest(beers, checkbox, (list, isChecked) -> {
                    if (isChecked) {
                        List<Beer> newList = Lists.newArrayList(list);
                        Collections.sort(newList);
                        return newList;
                    }
                    return list;
                });

        subscriptions.add(finalList.subscribe(subject::onNext));
    }

    @Override public Observable<List<Beer>> getBeers() {
        return subject.asObservable();
    }

    @Override public void unsubscribe() {
        if (subscriptions.hasSubscriptions())
            subscriptions.clear();
    }
}

File: app/src/main/java/com/jdappel/beerinvestigator/ui/viewmodel/impl/BeerViewModelImpl.kt
package com.jdappel.beerinvestigator.ui.viewmodel.impl

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import javax.inject.Inject
import com.jdappel.beerinvestigator.ui.viewmodel.BeerViewModel
import com.jdappel.beerinvestigator.data.repo.BreweryDBRepo
import com.jdappel.beerinvestigator.data.network.Result
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

/**
 * Implements [com.jdappel.beerinvestigator.ui.viewmodel.BeerViewModel] to retrieve a list of
 * beers based on the search criteria and populate the view.
 */
internal class BeerViewModelImpl @Inject constructor(private val beerService: BreweryDBRepo) :
    ViewModel(), BeerViewModel {
    private val subject = MutableLiveData<List<Beer>>()
    override fun subscribe(searchString: Flow<String>, checkbox: Flow<Boolean>) {
        viewModelScope.launch(Dispatchers.IO) {
            val searchResults: Flow<List<Beer>> = searchString.flatMapMerge { query: String ->
                beerService.findBeers(query)
                    .map { result ->
                        when(result) {
                            is Result.Success -> result.data!!
                            else -> listOf()
                        }                        }
            }
            val pair = searchResults.combine(checkbox) { first, second -> first to second}
            pair.collect {
                    if (it.second) {
                        subject.postValue(it.first.sorted())
                    }
                    subject.postValue(it.first)
                }
            }
    }

    override val beers: LiveData<List<Beer>>
        get() = subject
}